server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java

```diff
@@ -49,6 +49,15 @@ public void upsert(DbSession dbSession, BranchDto dto) {
     }
   }
 
+  public void updateMainBranchName(DbSession dbSession, String projectUuid, String newBranchKey) {
```

Comment:
same thing is done in ComponentKeyUpdaterDao. I can move the check to the WS but IMO upsert should not update the key. Generic methods to update a row should not change the ID of the row (and currently it doesn't). It's better to have a dedicated method for it.